standards/docker_standards.md

```diff
@@ -0,0 +1,230 @@
+EXPOSE ${PORT} 9229 9230
+USER node
+WORKDIR /home/node
+COPY --from=base --chown=node:node /home/node/package*.json ./
+COPY --from=base --chown=node:node /home/node/node_modules ./node_modules
+RUN npm install
+COPY --chown=node:node app/ ./app/
+RUN npm run build
+CMD [ "npm", "run", "start:watch" ]
+
+# Test stage copied in Jest configuration and declares the test task as the default command
+FROM development AS test
+USER node
+WORKDIR /home/node
+COPY --chown=node:node jest.config.js ./jest.config.js
+COPY --chown=node:node test/ ./test/
+CMD [ "npm", "run", "test" ]
+
+# Production stage exposes service port, copies in built app code and declares the Node app as the default command
+FROM base AS production
+USER node
+WORKDIR /home/node
+EXPOSE ${PORT}
+COPY --from=development /home/node/app/ ./app/
+CMD [ "node", "app/index" ]
+```
+
+## Docker Compose
+Docker compose can be used for local development and within a CI/CD pipeline to build docker containers.
+
+### Use override files to reduce duplication
+Additional settings can be applied to a docker compose file by using override files.
+
+Override files can be applied by listing the files after the `docker-compose` command with the `-f` parameter, i.e.
+
+`docker-compose up -f docker-compose.yaml -f docker-compose.override.yaml`
+
+Note that the above is equivalent to running the command:
+
+`docker-compose up`
+
+as calling `docker-compose` without specifying any files will run `docker-compose` with any available `docker-compose.yaml` and `docker-compose.override.yaml` files in the executing directory. 
+
+Note however that:
+
+`docker-compose up -f docker-compose.yaml`
+
+will **not** apply the docker `docker-compose.override.yaml` file, only the file specified.
+
+One use case is for testing - common settings can be put into the base `docker-compose.yaml` file, while changes to the command and container names can be placed in override files.
+
+The below example demonstrates changing the command and container name for testing:
+
+`docker-compose.yaml`
+
+```
+version: '3.4'
+services:
+  ffc-demo-service:
+    build: .
+    image: ffc-demo-service
+    container_name: ffc-demo-service
+    environment:
+      DEMO_API: http://demo-api
+
+volumes:
+  node_modules: {}
+
+```
+
+`docker-compose.test.yaml`
+```
+version: '3.4'
+services:
+  ffc-demo-service:
+    command: npm run test
+    container_name: ffc-demo-service-test
+```
+
+The tests can be run by providing the `docker-compose.test.yaml` file with a `-f` parameter:
+
+`docker-compose up -f docker-compose.yaml -f docker-compose.test.yaml`
+
+Further documentation on docker-compose can be found at https://docs.docker.com/compose/reference/overview/#specifying-multiple-compose-files.
+
+## Use projects to provide unique volumes and networks
+To avoid conflicts when running different permutations of docker files, projects should be specified to segregate the volumes and networks.
```

Comment:
Surely this just _adds_ extra permutations? :smile: 

So now, in order to verify why something "works on my machine" but doesn't work on yours I have to ask:

- What `Dockerfile` have you got?
- Which multi-stage build are you using?
- Which `docker-compose-XX.yaml` files have you included?
- What project name are you using?

It feels like we're creating a bit of a rod for our own backs.

Or can we maybe just simplify this down to a single standard setup for dev and a single standard setup for CI?

Reply:
I think it's when it's written down it appears more complex than it actually is.  Everyone will be using the same stage for the same task.

Want to develop, then use the develop stage
Want to test, then use the test stage
Want to mirror production, then use the production stage

Docker compose just targets those stages for you to make your docker commands shorter as well as bringing up whatever dependencies you need.  eg. development could also create a container with a Postgres database or a message queue.

The project name is more for CI/CD.  E.g. Jenkins will fail if multiple builds run simultaneous for a repo if they don't have a unique project name.  It can be an issue locally too of course but less likely.